materialize-slack/main.go

```diff
@@ -0,0 +1,357 @@
+			document[propName] = value
+		}
+	}
+
+	// Add the non-keys to the document.
+	for i, value := range values {
+		if i < len(b.spec.FieldSelection.Values) {
+			var propName = b.spec.FieldSelection.Values[i]
+
+			if raw, ok := value.([]byte); ok {
+				document[propName] = json.RawMessage(raw)
+			} else {
+				document[propName] = value
+			}
+		}
+	}
+	return document
+}
+
+func (d *transactor) Load(it *pm.LoadIterator, loaded func(int, json.RawMessage) error) error {
+	for it.Next() {
+		panic("Load should never be called for materialize-slack")
+	}
+	return nil
+}
+
+func (t *transactor) Store(it *pm.StoreIterator) (pm.StartCommitFunc, error) {
+	log.Debug("handling Store operation")
+
+	for it.Next() {
+		var b = t.bindings[it.Binding]
+		var parsed = buildDocument(b, it.Key, it.Values)
+		var tsStr, tsOk = parsed["ts"].(string)
+		var text, textOk = parsed["text"].(string)
+		var blocks, blocksOk = parsed["blocks"].(json.RawMessage)
+
+		if !tsOk {
+			return nil, fmt.Errorf("missing timestamp")
+		}
+
+		if !textOk {
+			return nil, fmt.Errorf("missing text")
+		}
+
+		// Parse the timstamp as a time.Time
+		ts, err := time.Parse(time.RFC3339Nano, tsStr)
+		if err != nil {
+			return nil, fmt.Errorf("invalid timestamp %q", tsStr)
+		}
+
+		var blocksParsed slack.Blocks
+		if blocksOk {
+			// Is this jank? This should really be taken care of by the
+			// serialization logic in slack.Block, but it can't be _that_
+			// bad to do it here... right? see below:
+			// https://api.slack.com/reference/surfaces/formatting#escaping
+			var escaped_blocks = string(blocks)
+			escaped_blocks = strings.ReplaceAll(escaped_blocks, "&", "&amp;")
+			escaped_blocks = strings.ReplaceAll(escaped_blocks, "<", "&lt;")
+			escaped_blocks = strings.ReplaceAll(escaped_blocks, ">", "&gt;")
+
+			err = json.Unmarshal([]byte(escaped_blocks), &blocksParsed)
+
+			if err != nil {
+				return nil, fmt.Errorf("invalid blocks value %q, %q: %w", reflect.TypeOf(blocks), string(blocks), err)
+			}
+		} else {
+			log.Warn(fmt.Sprintf("Blocks apparently NOT okay: %+v", reflect.TypeOf(parsed["blocks"])))
+		}
+
+		// Accept messages from at most 30 minutes in the past
+		if time.Since(ts).Minutes() < 10 {
+			if err := t.api.PostMessage(b.resource.Channel, text, blocksParsed.BlockSet, b.resource.SenderConfig); err != nil {
+				serializedBlocks, marshal_err := json.Marshal(blocksParsed)
+				if marshal_err == nil {
+					log.Warn(fmt.Sprintf("Parse Blocks: %+v", string(serializedBlocks)))
+				}
+				log.Warn(fmt.Errorf("error sending message: %w", err))
+			}
+			// Mom can we get rate limiting?
+			// we have rate limiting at home
+			// rate limiting at home:
+			time.Sleep(time.Second * 10)
+		} else {
+			log.Warn(fmt.Sprintf("Ignoring message from the past: %q", ts))
+		}
+	}
+
+	return func(ctx context.Context, runtimeCheckpoint *protocol.Checkpoint, runtimeAckCh <-chan struct{}) (*pf.ConnectorState, pf.OpFuture) {
```

Comment:
I don't think we need to have this be an async commit - see [materialize-google-pubsub](https://github.com/estuary/connectors/blob/60728e31593cc7c9a3d78c961b47eab3e84cca07/materialize-google-pubsub/transactor.go#L28-L63) for an example of a simpler store implementation that I think would work well for this.

Reply:
You're saying that the slack materialization connector doesn't need checkpoints, which makes sense now that we don't need to keep track of the last message timestamp and just accept any messages from the last 10 minutes. Removed it!